### PR TITLE
implement to find graphql tags to recognize a Relay file

### PR DESCRIPTION
**Pull request by asiandrummer**
Instead of using the incomplete babylon parser I wrote, use the one RelayCompiler is using to find the GraphQL tags.

During testing, I found a couple of bugs that pertained to the `textDocumentCache`, as well as the IDE lifecycle bug that couldn't find the handler for `shutdown` request. Fixed them by making them spec-compliant.

### CODE REVIEW

File: packages/graphql-language-service-server/src/findGraphQLTags.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        return;
+      }
+      const fragments = node.arguments[1];
+      if (fragments.type === 'ObjectExpression') {
+        fragments.properties.forEach(property => {
+          const tagName = getGraphQLTagName(property.value.tag);
+          const template = getGraphQLText(property.value.quasi);
+          if (tagName === 'graphql' || tagName === 'graphql.experimental') {
+            const loc = property.value.loc;
+            const range = new Range(
```
Comment (asiandrummer) [inline packages/graphql-language-service-server/src/findGraphQLTags.js]: This is where I compute the range of the quasi - needed in the language service to locate the query string and compute the range for, i.e. the range of diagnostic results.

File: packages/graphql-language-service-server/src/findGraphQLTags.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+      for (let ii = 2; ii < node.arguments.length; ii++) {
+        visit(node.arguments[ii], visitors);
+      }
+    },
+    TaggedTemplateExpression: node => {
+      const tagName = getGraphQLTagName(node.tag);
+      if (tagName != null) {
+        if (tagName === 'graphql' || tagName === 'graphql.experimental') {
+          const loc = node.loc;
+          const range = new Range(
```
Comment (asiandrummer) [inline packages/graphql-language-service-server/src/findGraphQLTags.js]: ^

File: packages/graphql-language-service-server/src/findGraphQLTags.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+              range,
+            });
+          }
+        });
+      } else {
+        const tagName = getGraphQLTagName(fragments.tag);
+        const template = getGraphQLText(fragments.quasi);
+        if (tagName === 'graphql' || tagName === 'graphql.experimental') {
+          const loc = fragments.loc;
+          const range = new Range(
```
Comment (asiandrummer) [inline packages/graphql-language-service-server/src/findGraphQLTags.js]: ^

**Comment by wincent** [comment]
Discussed offline some concerns about baking Relay-specifics into this, and also copy-pasta. This is probably an ok intermediate state however. Thoughts for future:

- `findGraphQLTags` is copy-pasta and going to diverge from the Relay implementation (especially given @kassens plans to rewrite it to be faster): ideally it would be somewhere we could import it easily.
- Given that Relay modern apps only contain just vanilla GraphQL, there is nothing intrinsically special about them in terms of the support that they require (other than knowing how to load schema extensions); rather than hard-coding Relay support, we could:
  - Make this configurable via `.graphqlconfig` (pass in a filter module for finding tags).
  - Make it extensible but support Relay by default.
- In this way people could add support for other frameworks at low cost. As @asiandrummer said, if people could supply parser modules we could even support non-JS use cases of colocated template strings, which would be pretty neat.

**Pull request merged**